File: apps/hat/src/domain/shortcuts/accelerator.ts
/**
 * Normalização pura de atalhos de teclado (accelerators).
 *
 * Forma canônica: modificadores na ordem CommandOrControl, Super, Alt, Shift,
 * seguidos de exatamente uma tecla final ('CommandOrControl+Alt+Shift+F').
 * Espelhado em crates/hat-core/src/accelerator.rs via
 * fixtures/accelerator-cases.json — mantenha os dois lados em sincronia.
 */

export type Platform = 'darwin' | 'win32';

type CanonicalModifier = 'CommandOrControl' | 'Super' | 'Alt' | 'Shift';

const MODIFIER_ORDER: readonly CanonicalModifier[] = [
  'CommandOrControl',
  'Super',
  'Alt',
  'Shift',
];

const MODIFIER_ALIASES: Readonly<Record<string, CanonicalModifier>> = {
  commandorcontrol: 'CommandOrControl',
  cmdorctrl: 'CommandOrControl',
  cmd: 'CommandOrControl',
  command: 'CommandOrControl',
  ctrl: 'CommandOrControl',
  control: 'CommandOrControl',
  meta: 'CommandOrControl',
  super: 'Super',
  win: 'Super',
  windows: 'Super',
  alt: 'Alt',
  option: 'Alt',
  opt: 'Alt',
  shift: 'Shift',
};

const DARWIN_SYMBOLS: Readonly<Record<CanonicalModifier, string>> = {
  CommandOrControl: '⌘', // ⌘
  Super: '⌘', // ⌘ (Super é a tecla Command no macOS)
  Alt: '⌥', // ⌥
  Shift: '⇧', // ⇧
};

const WIN32_LABELS: Readonly<Record<CanonicalModifier, string>> = {
  CommandOrControl: 'Ctrl',
  Super: 'Win',
  Alt: 'Alt',
  Shift: 'Shift',
};

/** Codes de keydown que são os próprios modificadores — nunca tecla final. */
const MODIFIER_CODES = new Set([
  'ControlLeft',
  'ControlRight',
  'ShiftLeft',
  'ShiftRight',
  'AltLeft',
  'AltRight',
  'MetaLeft',
  'MetaRight',
]);

const KEY_CODE_RE = /^key([a-z])$/i;
const DIGIT_CODE_RE = /^digit([0-9])$/i;
const FUNCTION_KEY_RE = /^f([1-9]|1[0-9]|2[0-4])$/i;
const SINGLE_ALNUM_RE = /^[a-z0-9]$/i;

/** Normaliza um token de tecla final ('KeyF' → 'F', 'Digit1' → '1', 'f' → 'F'). */
function normalizeKeyToken(token: string): string | null {
  const keyMatch = KEY_CODE_RE.exec(token);
  if (keyMatch?.[1] !== undefined) {
    return keyMatch[1].toUpperCase();
  }
  const digitMatch = DIGIT_CODE_RE.exec(token);
  if (digitMatch?.[1] !== undefined) {
    return digitMatch[1];
  }
  const fnMatch = FUNCTION_KEY_RE.exec(token);
  if (fnMatch?.[1] !== undefined) {
    return `F${fnMatch[1]}`;
  }
  if (SINGLE_ALNUM_RE.test(token)) {
    return token.toUpperCase();
  }
  // Teclas nomeadas (Space, Escape, ...): capitaliza a primeira letra.
  return token.charAt(0).toUpperCase() + token.slice(1);
}

/**
 * Normaliza um binding para a forma canônica, ou null se inválido
 * (sem tecla final, mais de uma tecla, segmento vazio ou nenhum
 * modificador não-Shift).
 */
export function normalize(binding: string): string | null {
  const parts = binding.split('+').map((part) => part.trim());
  if (parts.some((part) => part === '')) {
    return null;
  }

  const modifiers = new Set<CanonicalModifier>();
  let key: string | null = null;

  for (const part of parts) {
    const modifier = MODIFIER_ALIASES[part.toLowerCase()];
    if (modifier !== undefined) {
      modifiers.add(modifier);
      continue;
    }
    if (key !== null) {
      return null; // duas teclas finais
    }
    key = normalizeKeyToken(part);
    if (key === null) {
      return null;
    }
  }

  if (key === null) {
    return null; // sem tecla final
  }

  const hasNonShiftModifier = MODIFIER_ORDER.some(
    (modifier) => modifier !== 'Shift' && modifiers.has(modifier)
  );
  if (!hasNonShiftModifier) {
    return null;
  }

  const ordered = MODIFIER_ORDER.filter((modifier) => modifiers.has(modifier));
  return [...ordered, key].join('+');
}

/** True se o binding normaliza para uma forma canônica válida. */
export function isValid(binding: string): boolean {
  return normalize(binding) !== null;
}

/**
 * Rótulo de exibição por plataforma: símbolos colados no darwin ('⌘⇧F'),
 * nomes separados por '+' no win32 ('Ctrl+Shift+F').
 * Retorna string vazia para binding inválido.
 */
export function displayLabel(binding: string, platform: Platform): string {
  const canonical = normalize(binding);
  if (canonical === null) {
    return '';
  }

  const parts = canonical.split('+');
  const key = parts.pop() ?? '';
  const modifiers = parts as CanonicalModifier[];

  if (platform === 'darwin') {
    return modifiers.map((modifier) => DARWIN_SYMBOLS[modifier]).join('') + key;
  }
  return [...modifiers.map((modifier) => WIN32_LABELS[modifier]), key].join('+');
}

export interface AcceleratorKeyboardEvent {
  code: string;
  ctrlKey: boolean;
  metaKey: boolean;
  altKey: boolean;
  shiftKey: boolean;
}

/**
 * Constrói um binding canônico a partir de um KeyboardEvent (usa event.code
 * para paridade de layout físico). No darwin, metaKey vira CommandOrControl
 * e ctrlKey é ignorado (Control puro do macOS não tem representação
 * cross-platform); no win32, ctrlKey vira CommandOrControl e metaKey vira
 * Super. Retorna null se não formar binding válido.
 */
export function fromKeyboardEvent(
  event: AcceleratorKeyboardEvent,
  platform: Platform
): string | null {
  if (MODIFIER_CODES.has(event.code)) {
    return null; // keydown do próprio modificador, sem tecla final
  }

  const modifiers: CanonicalModifier[] = [];
  if (platform === 'darwin') {
    if (event.metaKey) {
      modifiers.push('CommandOrControl');
    }
  } else {
    if (event.ctrlKey) {
      modifiers.push('CommandOrControl');
    }
    if (event.metaKey) {
      modifiers.push('Super');
    }
  }
  if (event.altKey) {
    modifiers.push('Alt');
  }
  if (event.shiftKey) {
    modifiers.push('Shift');
  }

  const key = normalizeKeyToken(event.code);
  if (key === null) {
    return null;
  }

  return normalize([...modifiers, key].join('+'));
}

File: apps/hat/src-tauri/crates/hat-core/src/accelerator.rs
//! Normalização de atalhos. Forma canônica: `CommandOrControl+Alt+Shift+K`
//! (modificadores na ordem CommandOrControl, Super, Alt, Shift). Paridade
//! garantida com src/domain/shortcuts/accelerator.ts via fixture compartilhada
//! (fixtures/accelerator-cases.json).

/// Normaliza um binding para a forma canônica. Retorna None se inválido
/// (sem tecla final ou sem modificador não-Shift).
pub fn normalize(binding: &str) -> Option<String> {
    let parts: Vec<&str> = binding
        .split('+')
        .map(str::trim)
        .filter(|p| !p.is_empty())
        .collect();
    if parts.is_empty() {
        return None;
    }

    let mut cmd_or_ctrl = false;
    let mut super_mod = false;
    let mut alt = false;
    let mut shift = false;
    let mut key: Option<String> = None;

    for part in &parts {
        match part.to_ascii_lowercase().as_str() {
            "commandorcontrol" | "cmdorctrl" | "cmd" | "command" | "ctrl" | "control"
            | "meta" => cmd_or_ctrl = true,
            "super" | "win" | "windows" => super_mod = true,
            "alt" | "option" | "opt" => alt = true,
            "shift" => shift = true,
            _ => {
                if key.is_some() {
                    return None; // duas teclas finais
                }
                key = Some(normalize_key(part)?);
            }
        }
    }

    let key = key?;
    if !cmd_or_ctrl && !super_mod && !alt {
        return None; // exige ao menos um modificador não-Shift
    }

    let mut out: Vec<&str> = Vec::new();
    if cmd_or_ctrl {
        out.push("CommandOrControl");
    }
    if super_mod {
        out.push("Super");
    }
    if alt {
        out.push("Alt");
    }
    if shift {
        out.push("Shift");
    }
    out.push(&key);
    Some(out.join("+"))
}

pub fn is_valid(binding: &str) -> bool {
    normalize(binding).is_some()
}

/// Converte um W3C `KeyboardEvent.code` (ou tecla solta) em tecla canônica:
/// `KeyF` → `F`, `Digit1` → `1`, `f` → `F`, `F5` → `F5`.
fn normalize_key(raw: &str) -> Option<String> {
    let raw = raw.trim();
    if raw.is_empty() {
        return None;
    }
    if let Some(rest) = raw.strip_prefix("Key").filter(|r| r.len() == 1) {
        return Some(rest.to_ascii_uppercase());
    }
    if let Some(rest) = raw.strip_prefix("Digit").filter(|r| r.len() == 1) {
        return Some(rest.to_string());
    }
    // Teclas de função F1..F24 e nomeadas (Space, Escape…) passam como estão,
    // com inicial maiúscula; letras soltas viram maiúsculas.
    if raw.len() == 1 {
        return Some(raw.to_ascii_uppercase());
    }
    let mut chars = raw.chars();
    let first = chars.next()?.to_ascii_uppercase();
    Some(format!("{}{}", first, chars.as_str()))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn fixture() -> serde_json::Value {
        let raw = include_str!("../../../../fixtures/accelerator-cases.json");
        serde_json::from_str(raw).unwrap()
    }

    #[test]
    fn normalize_matches_shared_fixture() {
        let fx = fixture();
        let cases = fx["normalize"].as_array().unwrap();
        assert!(!cases.is_empty());
        for case in cases {
            let input = case["input"].as_str().unwrap();
            let expected = case["expected"].as_str().unwrap();
            assert_eq!(
                normalize(input).as_deref(),
                Some(expected),
                "normalize({:?})",
                input
            );
        }
    }

    #[test]
    fn validate_matches_shared_fixture() {
        let fx = fixture();
        let cases = fx["validate"].as_array().unwrap();
        assert!(!cases.is_empty());
        for case in cases {
            let input = case["input"].as_str().unwrap();
            let valid = case["valid"].as_bool().unwrap();
            assert_eq!(is_valid(input), valid, "is_valid({:?})", input);
        }
    }
}
